api/src/main/java/org/pmiops/workbench/actionaudit/ActionAuditQueryServiceImpl.java

```diff
@@ -0,0 +1,158 @@
+package org.pmiops.workbench.actionaudit;
+
+import com.google.cloud.bigquery.FieldValue;
+import com.google.cloud.bigquery.FieldValueList;
+import com.google.cloud.bigquery.QueryJobConfiguration;
+import com.google.cloud.bigquery.TableResult;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.function.Function;
+import java.util.stream.StreamSupport;
+import javax.annotation.Nullable;
+import javax.inject.Provider;
+import org.joda.time.Instant;
+import org.pmiops.workbench.api.BigQueryService;
+import org.pmiops.workbench.config.WorkbenchConfig;
+import org.pmiops.workbench.config.WorkbenchConfig.ActionAuditConfig;
+import org.pmiops.workbench.model.AuditLogEntriesResponse;
+import org.pmiops.workbench.model.AuditLogEntry;
+import org.pmiops.workbench.workspaces.WorkspaceService;
+import org.springframework.stereotype.Service;
+
+@Service
+public class ActionAuditQueryServiceImpl implements ActionAuditQueryService {
+
+  private BigQueryService bigQueryService;
+  private Provider<WorkbenchConfig> workbenchConfigProvider;
+  private WorkspaceService workspaceService;
+  public static final String WORKSPACE_EVENTS_QUERY_STRING_FORMAT =
```

Comment:
I'm really close to shopping for a template engine for this. I'm already up to three ordered substitutions and I don't have any time constraints, ordering options, etc.

Reply:
Maybe look to https://cloud.google.com/bigquery/docs/parameterized-queries ? I think we use this on the CB side of the house, so there should be examples in our codebase ([like this](https://github.com/all-of-us/workbench/blob/49b293e10583d5bbb047daf87f70bc6d45aa24cf/api/src/main/java/org/pmiops/workbench/cdr/ConceptBigQueryService.java#L57-L58)).